File: src/main/scalacode/com/huangning/scala/lesson/collection/ListNote.scala
package com.huangning.scala.lesson.collection

import scala.collection.mutable.ListBuffer

/**
 * Author: 黄宁
 * Date: 2020/06/26
 * Target: 了解List（列表）的基本用法
 * Note:
 * 1. List如果不写泛型，有时候编译器推断不出来类型，并显示nothing
 * 2. 写泛型的好处就是如果写错成其他类型，会有报错提示
 */
object ListNote {
  def main(args: Array[String]): Unit = {

    // 创建List, Nil是长度为0的list
    val l1: List[Int] = List[Int](1, 2, 3, 4, 5)
    val l2: List[String] = List[String]("hello scala", "hello spark", "hello java", "hn", "tony")

    // 两种遍历方式
    for (elem <- l1) println(elem)
    l1.foreach(println)

    /**
     * List的方法：
     * 1. filter: 将返回是true的元素保留下来
     * 2. count: 计算符合条件的元素个数
     * 3. map: 一一映射
     * 4. flatMap: 先map，再flat
     */
    val res1: List[String] = l2.filter(s => {
      "hello scala".equals(s)
    })

    val res2: Int = l2.count(s => {
      s.length >= 4
    })

    val res3: List[Array[String]] = l2.map(elem => {
      elem.split(" ")
    })

    val res4: List[String] = l2.flatMap(elem => {
      elem.split(" ")
    })

    // 创建可变长的List
    val list: ListBuffer[Int] = ListBuffer[Int](1, 2, 3, 4)
    list.append(4, 5, 6) // 后向追加多个元素
    list.+=(1000, 2000) // 后向追加多个元素
    list.+=:(100) // 前向追加多个元素

  }
}

File: src/main/javacode/com/huangning/java/lesson/collection/ListNote.java
package com.huangning.java.lesson.collection;

import java.util.Collections;
import java.util.LinkedList;
import java.util.List;

/**
 * Author: 黄宁
 * Date: 2021/02/17
 * Func: 了解List
 */
public class ListNote {
    public static void main(String[] args) {

        java.util.List<String> l1 = new LinkedList<>();

        // 传统的遍历
        for(int i = 0; i <= 5; i++){
            l1.add("a-" + i);
        }

        // 增强for循环遍历
        for(String elem : l1){
            l1.add("a-" + elem);
        }

        // 添加元素
        l1.add(3, "a1");

        // 获取元素
        System.out.print("索引为2的元素: " + l1.get(2));

        // 更新元素
        String old = l1.set(6, "a200");
        System.out.println("被替换的旧元素: " + old);

        // 元素第一次出现的索引
        System.out.println("第一次出现'a-3'对应的索引为：" + l1.indexOf("a3"));

        // 元素最后一次出现的索引
        System.out.println("最后一次出现'a3'对应的索引为：" + l1.lastIndexOf("a3"));

        // 删除第一次出现的某个元素
        l1.remove("a-1");
        // 删除索引为3的元素
        l1.remove(3);

        // Collections常用方法
        Collections.shuffle(l1);
        System.out.println("随机排列后的类l1: " + l1);

        Collections.reverse(l1);
        System.out.println("反转后的l1为: " + l1);

        Collections.sort(l1);
        System.out.println("排序后的l1为: " + l1);

        System.out.println("二分法查找'a2'对应的索引为: " + Collections.binarySearch(l1,"a2"));

    }
}
